### PR TITLE
[SPARK-26111][ML][WIP] Support F-value between label/feature for continuous distribution feature selection

### PR DESCRIPTION
**Pull request by huaxingao**
### What changes were proposed in this pull request?
Add FValue selection for continuous distribution features.

### Why are the changes needed?
Current Spark only supports the selection of categorical features, while there are many requirements for the selection of continuous distribution features.

ANOVA F-value is one way to select features from the continuous distribution and it's important to support it in spark.

### Does this PR introduce any user-facing change?
Yes

### How was this patch tested?
Add unit tests

### CODE REVIEW

**Comment by huaxingao** [comment]
@srowen 
Hi Sean, I am thinking of adding selector for continuous distribution features. I want to ask your opinion before I go any further. I will also ask Ruifeng after Chinese New Year holiday. I bet he is on vacation so don't want to ping him now. 

Currently, Spark only supports selection of categorical features (```ChiSqSelector```). I am thinking of adding two new selectors for continuous distribution features:
1. ```FValueRegressionSelector``` for  continuous features and continuous labels.
2. ```FValueClassificationSelector``` for  continuous features and categorical labels.

Currently, this WIP PR only has ```FValueRegressionSelector``` implemented. ```FValueClassificationSelector``` is very similar. The calculation for classification f value is a little more complicated. I wrote the pseudo code here along with an example:

```
  // pseudo code:
  // for each feature:
  //     separate feature values into array of arrays by label (call it arr)
  //       e.g. if feature is [3.3, 2.5, 1.0, 3.0, 2.0] and labels are [1, 2, 1, 3, 3]
  //       then output should be arr = [[3.3, 1.0], [2.5], [3.0, 2.0]]
  //     n_classes = len(arr) (num. of distinct label categories)
  //     n_samples_per_class = [len(a) for a in arr]
  //     n_samples = sum(n_samples_per_class)    (= num. of rows in feature column)
  //       e.g. in above example, n_classes = 3, n_samples_per_class = [2, 1, 2], n_samples = 5
  //     ss_all = sum of squares of all in feature (e.g. 3.3^2+2.5^2+1.0^2+3.0^2+2.0^2)
  //     sq_sum_all = square of sum of all data (e.g. (3.3+2.5+1.0+3.0+2.0)^2)
  //     sq_sum_classes = [sum(a) ** 2 for a in arr]  (e.g. [(3.3+1.0)^2, 2.5^2, (3.0+2.0)^2]
  //     sstot = ss_all - (sq_sum_all / n_samples)
  //     ssbn = sum( sq_sum_classes[k] / n_samples_per_class[k] for k in range(n_classes)) - (sq_sum_all / n_samples)
  //       e.g. ((3.3+1.0)^2 / 2 + 2.5^2 / 1 + (3.0+2.0)^2 / 2) - sq_sum_all / 5
  //     sswn = sstot - ssbn
  //     dfbn = n_classes - 1
  //     dfwn = n_samples - n_classes
  //     msb = ssbn / dfbn
  //     msw = sswn / dfwn
  //     f = msb / msw
  //     pvalue = 1 - FDistribution(dfbn, dfwn).cdf(f)
```
sklean has both f_regression and f_classif. Here are the links:
https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_regression.html#sklearn.feature_selection.f_regression
https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.f_classif.html#sklearn.feature_selection.f_classif

**Comment by srowen** [comment]
I think it's fairly reasonable, yeah. F-value is pretty standard and not hard to compute.

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+  @Since("3.1.0")
+  override def load(path: String): FRegressionSelector = super.load(path)
+}
+
+/**
+ * Model fitted by [[FRegressionSelector]].
+ */
+@Since("3.1.0")
+final class FRegressionSelectorModel private[ml] (
+    @Since("3.1.0") override val uid: String,
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: I found that `f_regression` in scikit-learn will return both arrays of F-values and P-values, can we also add them to  `FRegressionSelectorModel`?

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+   @Since("3.1.0")
+   override def write: MLWriter = null // new FRegressionSelectorModelWriter(this)
+
+  @Since("3.1.0")
+  override def toString: String = {
+    s"FRegressionSelectorModel: uid=$uid, numSelectedFeatures=${selectedFeatures.length}"
+  }
+
+  private[spark] def compressSparse(indices: Array[Int],
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: lint

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    var i = 0
+    var j = 0
+    var indicesIdx = 0
+    var filterIndicesIdx = 0
+    while (i < indices.length && j < filterIndices.length) {
+      indicesIdx = indices(i)
+      filterIndicesIdx = filterIndices(j)
+      if (indicesIdx == filterIndicesIdx) {
+        newIndices += j
+        newValues += values(i)
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]:
```
val value = values(i)
if (value != 0) {
...
}
```

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+  override def load(path: String): FRegressionSelector = super.load(path)
+}
+
+/**
+ * Model fitted by [[FRegressionSelector]].
+ */
+@Since("3.1.0")
+final class FRegressionSelectorModel private[ml] (
+    @Since("3.1.0") override val uid: String,
+    val selectedFeatures: Array[Int])
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: add a requirement to make sure that `selectedFeatures` is sorted?
```scala
    var prev = -1
    selectedFeatures.foreach { i =>
      require(prev < i, s"Index $i follows $prev and is not strictly increasing")
      prev = i
    }
```

File: mllib/src/main/scala/org/apache/spark/ml/stat/FRegressionTest.scala [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    val rdd = dataset.select(col(labelCol).cast("double"), col(featuresCol)).as[(Double, Vector)]
+      .rdd.map { case (label, features) => LabeledPoint(label, features) }
+
+    val numOfFeatures = rdd.first().features.size
+    val numOfSamples = rdd.count()
+    val degreeOfFreedom = numOfSamples.toInt - 2
+
+    var fTestResultArray = new Array[FRegressionTestResult](numOfFeatures)
+    val labels = rdd.map(d => d.label)
+    for (i <- 0 until numOfFeatures) {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/stat/FRegressionTest.scala]: compute each col at once?
This should be inefficient, I guess only one pass is needed.

File: mllib/src/main/scala/org/apache/spark/ml/stat/FRegressionTest.scala [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      fValue: Double)
+
+  /**
+   * @param dataset  DataFrame of continuous labels and continuous features.
+   * @param featuresCol  Name of features column in dataset, of type `Vector` (`VectorUDT`)
+   * @param labelCol  Name of label column in dataset, of any numerical type
+   * @return Array containing the FRegressionTestResult for every feature against the label.
+   */
+  @Since("3.1.0")
+  def test_regression(dataset: Dataset[_], featuresCol: String, labelCol: String):
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/stat/FRegressionTest.scala]: this method name `test_regression` should follow Camel-Case

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+        } else {
+          i += 1
+        }
+      }
+    }
+    // TODO: Sparse representation might be ineffective if (newSize ~= newValues.size)
+    (newIndices.result(), newValues.result())
+  }
+
+  private[spark] def compressDense(values: Array[Double]): Array[Double] = {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: the logic here is simple, just put it in the above defination of `func`

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala (hunk 3 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: here requires `selectedFeatures` in model is sorted?

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+        FTestResult
+          .sortBy { case (res, _) => res.pValue }
+          .take((FTestResult.length * getPercentile).toInt)
+      case "fpr" =>
+        FTestResult
+          .filter { case (res, _) => res.pValue < getFpr }
+      case errorType =>
+        throw new IllegalStateException(s"Unknown FRegressionSelector Type: $errorType")
+    }
+    val indices = features.map { case (_, index) => index }
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: it seems that this `indices` need to be sorted?

**Comment by zhengruifeng** [comment]
> Currently, this WIP PR only has FValueRegressionSelector implemented. FValueClassificationSelector is very similar. The calculation for classification f value is a little more complicated.

I think `f_classif` is different enough for another PR.

File: mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala (hunk 2 above)
Comment (huaxingao) [reply mllib/src/main/scala/org/apache/spark/ml/feature/FRegressionSelector.scala]: Sorry I am not sure what you want me to do for this one and the next one.  Could you please clarify?

**Comment by huaxingao** [comment]
I will close this WIP and submit  a new PR. 

@zhengruifeng 
I will address your comments in the new PR.
